pwndbg/dbg/lldb/__init__.py

```diff
@@ -304,6 +698,314 @@ def vmmap(self) -> pwndbg.dbg_mod.MemoryMap:
 
         return LLDBMemoryMap(pages)
 
+    def find_largest_range_len(
+        self, min_search: int, max_search: int, test: Callable[[int], bool]
+    ) -> int:
+        """
+        Finds the largest memory range given a minimum and a maximum value
+        for the size of the rage. This is a binary search, so it should do on
+        the order of log2(max_search - min_search) attempts before it arrives at
+        an answer.
+        """
+        # See if there's even any region we could possibly read.
+        r = max_search - min_search
+        if r == 0:
+            return min_search if test(min_search) else 0
+
+        # Pick the midpoint from our previous search.
+        mid_search = min_search + r // 2
+
+        if not test(mid_search):
+            # No dice. This means the limit of the mapping must come before the
+            # midpoint.
+            return self.find_largest_range_len(min_search, mid_search, test)
+
+        # We can read this range. This means that the limit of the mapping must
+        # come after the midpoint, or be equal to it, exactly.
+        after = self.find_largest_range_len(mid_search + 1, max_search, test)
+        if after > 0:
+            # It came after the midpoint.
+            return after
+
+        # We are exactly at the limit.
+        return min_search
+
+    @override
+    def read_memory(self, address: int, size: int, partial: bool = False) -> bytearray:
+        if size == 0:
+            return bytearray()
+
+        # Try to read exactly the requested size.
+        e = lldb.SBError()
+        buffer = self.process.ReadMemory(address, size, e)
+        if buffer:
+            return buffer
+        elif not partial:
+            raise pwndbg.dbg_mod.Error(f"could not read {size:#x} bytes: {e}")
+
+        # At this point, we're in a bit of a pickle. LLDB doesn't give us enough
+        # information to find out what the last address it can read from is. For
+        # all we know, it could be any address in the range (address, address+size),
+        # so we have to get creative.
+        #
+        # First, try to derive that information from the mmap.
+        first_page = None
+        last_page = None
+        vmmap_size = 0
+        for page in self.vmmap().ranges():
```

Comment:
lldb support `GetMemoryRegionContainingAddress` do you think it is worth using it here?

Reply:
`GetMemoryRegionContainingAddress` afaik doesn't give us the index of the page in the `SBMemoryRegionInfoList`, just the page itself. Since we have to scan a range of pages starting at the page where the address is, it's not that useful to us here.